src/main/java/org/opentripplanner/routing/algorithm/raptoradapter/transit/request/PriorityGroupConfigurator.java

```diff
@@ -0,0 +1,101 @@
+package org.opentripplanner.routing.algorithm.raptoradapter.transit.request;
+
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.stream.Stream;
+import org.opentripplanner.framework.lang.ArrayUtils;
+import org.opentripplanner.routing.algorithm.raptoradapter.transit.cost.grouppriority.TransitPriorityGroup32n;
+import org.opentripplanner.routing.api.request.request.filter.TransitPriorityGroupSelect;
+import org.opentripplanner.transit.model.framework.FeedScopedId;
+import org.opentripplanner.transit.model.network.RoutingTripPattern;
+
+/**
+ * This class dynamically builds an index of transit-group-ids from the
+ * provided {@link TransitPriorityGroupSelect}s while serving the caller with
+ * group-ids for each requested pattern. It is made for optimal
+ * performance, since it is used in request scope.
+ * <p>
+ * THIS CLASS IS NOT THREAD-SAFE.
+ */
+public class PriorityGroupConfigurator {
+
+  private static final int BASE_GROUP_ID = TransitPriorityGroup32n.groupId(0);
+  private int groupIndexCounter = 0;
+  private final boolean enabled;
+  private final PriorityGroupMatcher[] baseMatchers;
+  private final PriorityGroupMatcher[] agencyMatchers;
+  private final PriorityGroupMatcher[] globalMatchers;
+  private final Map<PriorityGroupMatcher, Map<FeedScopedId, Integer>> agencyMatchersIds = new HashMap<>();
+  private final Map<PriorityGroupMatcher, Integer> globalMatchersIds = new HashMap<>();
```

Comment:
Good catch! The performance is super fast for the default - it check if the memory ref is the same - does not even need to access the memory to do it. OTP uses this in some cases, but we do not want to do that here. I will have a look on how to fix this. But, I think just the standard way is what we should do. Also, the Java Map implementations cashes the hash, witch sometimes lead to surprising bugs if you mutate the key - you should not do it, but it has happened :-)

Reply:
After looking at this in more detail I refactored to avoid using hashCode and equals in the Matchers. Using the default would work in this case, but eliminating the maps is much better. I have now added unit-tests and fixed a bug as well.